lib/api/objectPutLegalHold.js

```diff
@@ -0,0 +1,124 @@
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+function objectPutLegalHold(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'objectPutLegalHold' });
+
+    const { bucketName, objectKey } = request;
+
+    const decodedVidResult = decodeVersionId(request.query);
+    if (decodedVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decodedVidResult,
+        });
+        return process.nextTick(() => callback(decodedVidResult));
+    }
+    const versionId = decodedVidResult;
+
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        objectKey,
+        requestType: 'objectPutLegalHold',
+        versionId,
+    };
+
+    return async.waterfall([
+        next => metadataValidateBucketAndObj(metadataValParams, log,
+        (err, bucket, objectMD) => {
+            if (err) {
+                log.trace('request authorization failed',
+                    { method: 'objectPutLegalHold', error: err });
+                return next(err);
+            }
+            if (!objectMD) {
+                const err = reqVersionId ? errors.NoSuchVersion :
+                    errors.NoSuchKey;
+                log.trace('error no object metadata found',
+                    { method: 'objectPutLegalHold', error: err });
+                return next(err, bucket);
+            }
+            if (objectMD.isDeleteMarker) {
+                log.trace('version is a delete marker',
+                    { method: 'objectPutLegalHold' });
+                return next(errors.MethodNotAllowed, bucket);
+            }
+            if (!bucket.isObjectLockEnabled()) {
+                log.trace('object lock not enabled on bucket',
+                    { method: 'objectPutLegalHold' });
+                return next(errors.InvalidRequest.customizeDescription(
+                    'Bucket is missing Object Lock Configuration'
+                ), bucket);
+            }
+            return next(null, bucket, objectMD);
+        }),
+        (bucket, objectMD, next) => {
+            log.trace('parsing legal hold');
+            parseLegalHoldXml(request.post, log, (err, res) =>
+                next(err, bucket, res, objectMD));
+        },
+        (bucket, legalHold, objectMD, next) => {
+            // eslint-disable-next-line no-param-reassign
+            objectMD.legalHold = legalHold;
+            const params = objectMD.versionId ?
+                { versionId: objectMD.versionId } : {};
+            const replicationInfo = getReplicationInfo(objectKey, bucket, true,
+                0, REPLICATION_ACTION, objectMD);
+            if (replicationInfo) {
+                // eslint-disable-next-line no-param-reassign
+                objectMD.replicationInfo = Object.assign({},
+                    objectMD.replicationInfo, replicationInfo);
```

Comment:
@dora-korpar In here I used similar logic with your PUT object retention PR to set the replication info but couldn't be sure if we'd necessarily need it for this one as well. What do you think?

Reply:
yeah, I borrowed that logic from the `putObjectTagging` api, but actually I don't see any reason to include it. I think we should both remove it from our apis